### ISSUE TITLE
Support declaring dependency constraint to specific commit id used in source dependencies

### ISSUE THREAD

**Issue by lacasseio** [gradle/gradle-native#883]
### Expected Behavior

The following should be allowed:
```
dependencies {
    implementation('org.gradle.cpp-samples:utilities') {
        version {
            commitId = 'sha-1-commit-id'
        }
    }
}
```
We could also use the version field and if we can't find the version in the list of tag, we try to find that specific string as a commit id, then fail.

### Current Behavior

Users can't add a source dependency to a specific commit id.

### Your Environment

 * Build scan URL:

**Comment by ljacomet** [comment]
Ideally this should use the version and not a new field.
Adding a new field would be very disruptive to the dependency resolution code I think.

In addition, using existing data would facilitate support of dependency locking for example.

**Comment by ppalaga** [comment]
If you choose to use the version, then I think making the version to contain the sha1 and nothing else will make it hard to follow what state of the dependency is behind that. 

When I see `org.mygroup:my-artifact:abcd1234` I have no idea whether it is new or old from the compatibility PoV, etc.

In srcdeps [1], users can put a prefix of their choice before the sha1. One typically puts a human readable version there, e.g. `1.2.3-SRC-revision-abcd1234` or some version stream marker `1.2.x-SRC-revision-abcd1234`. (`SRC-revision` is a marker for srcdeps to identify source dependencies - that's an unimportant detail for the current discussion.)

[1] https://github.com/srcdeps/srcdeps-maven

**Comment by bigdaz** [comment]
I'd like to see issues unrelated to `gradle/gradle-native` go into the `gradle/gradle` issue tracker. Having silos based on the team that did the implementation doesn't seem like a good practice.

**Comment by big-guy** [comment]
@bigdaz We use the board to organize whatever we're working on (so not strictly native things), but I agree that we don't want to lose visibility (or expect people to find issues on gradle-native for non-native things).  We (the native team) should take care to duplicate issues that need to be visible on gradle/gradle.

I have a hesitation to use gradle/gradle solely for these kinds of issues. It would just be a dumping ground we'd never revisit.
